ufl/cell.py

```diff
@@ -1,298 +1,456 @@
-"Types for representing a cell."
+"""Types for representing a cell."""
 
 # Copyright (C) 2008-2016 Martin Sandve Alnæs
 #
 # This file is part of UFL (https://www.fenicsproject.org)
 #
 # SPDX-License-Identifier:    LGPL-3.0-or-later
 
+from __future__ import annotations
 import functools
 import numbers
+import typing
 
-import ufl.cell
-from ufl.core.ufl_type import attach_operators_from_hash_data
+from ufl.core.ufl_type import UFLObject
+from abc import abstractmethod
+
+try:
+    from typing import Self
+except ImportError:
+    # Hack to get this to work pre Python 3.11
+    # Delete this after 04 Oct 2026 (Python 3.10 end of life)
+    # After this date, can also replace "self: Self," with "self,
+    from typing import TypeVar
+    Self = TypeVar("Self", bound="AbstractCell")
 
-# Export list for ufl.classes
 __all_classes__ = ["AbstractCell", "Cell", "TensorProductCell"]
 
 
-# --- The most abstract cell class, base class for other cell types
+class AbstractCell(UFLObject):
```

Comment:
My usual preferred approach these days for abstract interfaces is `typing.Protocol` rather than `ABC`. That said, I think because you have implementation in this class and third-party libraries will inherit from it, `Protocol` is not necessarily better.

WDYT?

Reply:
In the element class, I think it has to be an ABC for now, as there's some implementation in the base class.

Long term, we should maybe aim for `Protocol`s, but for now I think it's confusing to have a mixture of the two, so I'm leaning towards making them `ABC`s for now